packages/vite/src/node/debugger.ts

```diff
@@ -0,0 +1,52 @@
+import debug from 'debug'
+
+// set in bin/vite.js
+const filter = process.env.VITE_DEBUG_FILTER
+
+const DEBUG = process.env.DEBUG
+
+interface DebuggerOptions {
+  onlyWhenFocused?: boolean | string
+}
+
+export function createDebugger(
+  ns: string,
+  options: DebuggerOptions = {}
+): debug.Debugger['log'] {
+  const log = debug(ns)
+  const { onlyWhenFocused } = options
+  const focus = typeof onlyWhenFocused === 'string' ? onlyWhenFocused : ns
+  return (msg: string, ...args: any[]) => {
+    if (filter && !msg.includes(filter)) {
+      return
+    }
+    if (onlyWhenFocused && !DEBUG?.includes(focus)) {
+      return
+    }
+    log(msg, ...args)
+  }
+}
+
+/**
+ * Set DEBUG environment variable to turn on debugging
+ * E.g. DEBUG="vite:resolve"
+ * To see all debug logs set DEBUG="vite:*" though it will be noisy
+ */
+export const CODES = {
```

Comment:
**suggestion:1 (dx):** Could we find a better name for this?

**suggestion:2 (dx):** Maybe we should use `Object.freeze` here, so nothing gets added at runtime.

Reply:
done